### PR TITLE
fix: add scrollable area check and scrollable area attribute

### LINKED ISSUE #2733: Page scroll position not reset to top on navigation (regression)
### Describe the bug

When navigating to a new page, if you are already scrolled partway down the page, the next page load will keep the current scroll position and not scroll back to the top.

I believe this broke with next-193. When I install next-192, this does not happen -- navigating to a new page always scrolls to the top, just like regular browser navigation.

### Reproduction

Init the SvelteKit demo project. Add the following paragraph in `about.svelte` after the last paragraph.

```html
<p>Go <a href="/">home</a></p>
```

Make the browser window small enough so that the page scrolls and scroll to the bottom of the page. Click the Home link. The Home page loads, but does not scroll to the top (i.e. the nav bar is not visible).

scroll-bug.

Tested in Chrome and Firefox on Windows 10.

### Logs

_No response_

### System Info

```shell
System:
    OS: Linux 4.19 Ubuntu 20.04.1 LTS (Focal Fossa)
    CPU: (4) x64 Intel(R) Core(TM) i7-6600U CPU @ 2.60GHz
    Memory: 7.79 GB / 12.40 GB
    Container: Yes
    Shell: 5.0.17 - /bin/bash
  Binaries:
    Node: 14.17.0 - ~/.nvm/versions/node/v14.17.0/bin/node
    Yarn: 1.22.5 - /usr/bin/yarn
    npm: 8.1.1 - ~/.nvm/versions/node/v14.17.0/bin/npm
  Browsers:
    Chrome: 89.0.4389.90
  npmPackages:
    @sveltejs/kit: ^1.0.0-next.193 => 1.0.0-next.193
    svelte: ^3.34.0 => 3.44.1
```

### Severity

serious, but I can work around it

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by eltigerchino**
fixes #2733

Adds a check to see if `html` or `body` is scrollable.
Otherwise, it looks for an element with the attribute `data-sveltekit-main-scrollable` to determine which element to scroll during navigation.

Looking for feedback as this solution still feels quite rough. @dummdidumm 

#### TODO
- [ ] add tests
- [ ] docs
- [ ] changeset

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [ ] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### CODE REVIEW

File: packages/kit/src/runtime/client/client.js [hunk 1]
```diff
@@ -63,6 +63,62 @@ function update_scroll_positions(index) {
...
+
+const SCROLLABLE_AREA_ATTRIBUTE = 'data-sveltekit-main-scrollable';
+
+function get_scrollable_area() {
+	const html_is_scrollable = getComputedStyle(document.documentElement).overflowY !== 'hidden';
+	const scrollable_area = html_is_scrollable
+		? document.documentElement
+		: is_scrollable(document.body);
+
+	const custom_scrollable_areas = document.querySelectorAll(`*[${SCROLLABLE_AREA_ATTRIBUTE}]`);
```
Comment (dummdidumm) [inline packages/kit/src/runtime/client/client.js:84]: probably better to only do `querySelector` and move the `..All` inside the DEV check

**Review by dummdidumm — COMMENTED** [review]
The idea is tempting, although I'm wondering how much code people really save by doing that. I'm also wondering if they would always prefer the explicitly set scrollable area to be used for scrolling. Also, how often will be html element be non-scrollable while the body will be? Feels like both won't have that if someone does that.

**Comment by s3812497** [comment]
> The idea is tempting, although I'm wondering how much code people really save by doing that. I'm also wondering if they would always prefer the explicitly set scrollable area to be used for scrolling. Also, how often will be html element be non-scrollable while the body will be? Feels like both won't have that if someone does that.

I agree. The thread showed that many people were able to solve it with their own workarounds.
It might be enough to document the behaviour and/or provide a warning when `html` isn't scrollable.

The common patterns I've gathered from the issue thread are:

1. `html` is not scrollable and its `height` is set to the window height.
`body` is the scrolling area.
2. `html` and/or `body` are not scrollable and both have `height` set to window height.
`div` or `main` is the scrolling area. The element location varies.

**Comment by petem-methods** [comment]
I think warning + documenting is the most important thing - it is very non-obvious that this will happen, mainly because as a user you're not thinking about how the magic happens.

> The thread showed that many people were able to solve it with their own workarounds.

But I really like the idea of providing a proper solution - as a mortal Sveltekit user, I don't feel like I understand the workaround we have done enough to know that I've done it right. Like, I added the check for when `navigation.type === "link"`... because otherwise `goto` didn't work properly.... I want to be able to just say, "hey Sveltekit, treat *this* as the scrollable area" and have it work in the same way as it does for the html element. 

Please let me know if I can help testing this - and if so, how. 🙏

**Comment by s3812497** [comment]
> But I really like the idea of providing a proper solution - as a mortal Sveltekit user, I don't feel like I understand the workaround we have done enough to know that I've done it right. Like, I added the check for when `navigation.type === "link"`... because otherwise `goto` didn't work properly.... I want to be able to just say, "hey Sveltekit, treat _this_ as the scrollable area" and have it work in the same way as it does for the html element.

In this case, it would be preferable to always require the explicit scrolling area set. I hesitated earlier about this when I recalled a conversation about the issues of having too many framework specific attributes. 

> Please let me know if I can help testing this - and if so, how. 🙏

I had the idea to add tests for the new attribute following the procedure:
1. Scroll down.
2. Click a link.
3. Check scroll position of the main scrolling element.

However, I'm waiting for more feedback from everyone before continuing this PR.

**Comment by Rich-Harris** [comment]
I wonder if #8710 is an adequate solution to this, if properly documented? It's a tiny bit more work than adding an attribute, but it uses existing API surface area and teaches people a technique that they can then generalise to other problems.

**Comment by s3812497** [comment]
It would also need to address the issue of correctly scrolling to the top when navigating to a new page.

My current understanding of snapshots is that it only restores scroll when navigating backwards / forwards to visited pages.

**Comment by petem-methods** [comment]
It also doesn't stop the mistake being made, I don't think? We only discovered it late in a cycle when testing on real mobile devices, so the buggy scrolling went into production. So we need something like this `DEV` check / warn. 

As an aside, I'm not sure I've seen an explanation here of what the relevant difference is in mobile browsers that causes this behaviour to manifest (or - rather - *not* manifest reliably desktop). 🙏

**Comment by samal-rasmussen** [comment]
@Rich-Harris said:
> Snapshots are restored immediately after `afterNavigate` callbacks

Unfortunately I'm seeing afterNavigate being called after snapshot restore, so this doesn't work.

**Comment by dvlpr91** [comment]
I was very confused that this was not in the sveltekit docs.

**Pull request closed without merging**